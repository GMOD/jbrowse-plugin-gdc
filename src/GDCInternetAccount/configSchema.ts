import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { Instance } from 'mobx-state-tree'
import PluginManager from '@jbrowse/core/PluginManager'
import { createInternetAccountConfig } from '@jbrowse/core/pluggableElementTypes/models'

function GDCConfigFactory(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'GDCInternetAccount',
    {
      validDomains: {
        description:
          'array of valid domains the url can contain to use this account. Empty = all domains',
        type: 'stringArray',
        defaultValue: [],
      },
      authHeader: {
        description: 'custom auth header for authorization',
        type: 'string',
        defaultValue: 'Authorization',
      },
      customEndpoint: {
        description: 'custom endpoint for the external token resource',
        type: 'string',
        defaultValue: '',
      },
    },
    {
      baseConfiguration: createInternetAccountConfig(pluginManager),
      explicitlyTyped: true,
    },
  )
}

export type GDCInternetAccountConfigModel = ReturnType<typeof GDCConfigFactory>
export type OAuthInternetAccountConfig = Instance<GDCInternetAccountConfigModel>
export default GDCConfigFactory
