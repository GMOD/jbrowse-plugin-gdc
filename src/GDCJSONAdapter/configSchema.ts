import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

export default ConfigurationSchema(
  'GDCJSONAdapter',
  {
    featureType: {
      type: 'stringEnum',
      model: types.enumeration('Feature Type', ['mutation', 'gene']),
      defaultValue: 'mutation',
      description: 'The type of the track to add',
    },
    data: {
      type: 'string',
      defaultValue: '',
      description: 'JSON string of the GDC data from the file'
    }
  },
  { explicitlyTyped: true },
)
