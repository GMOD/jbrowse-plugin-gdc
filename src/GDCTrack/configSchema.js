import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export default pluginManager => {
    const { BaseTrackConfig: LinearGenomeTrackConfig }  = pluginManager.getPlugin('LinearGenomeViewPlugin').exports
    return ConfigurationSchema(
    'GDCTrack',
    {
      adapter: pluginManager.pluggableConfigSchemaType('adapter'),
      renderer: pluginManager.pluggableConfigSchemaType('renderer'),
    },
    { baseConfiguration: LinearGenomeTrackConfig, explicitlyTyped: true },
  )
}
