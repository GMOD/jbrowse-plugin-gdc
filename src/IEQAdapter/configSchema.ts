import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'IeqAdapter',
  {
    ieqLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/myfile.tsv', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
