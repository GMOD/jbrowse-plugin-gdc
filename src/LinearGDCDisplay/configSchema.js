import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default pluginManager => {
  const { baseLinearDisplayConfigSchema } = pluginManager.getPlugin(
    'LinearGenomeViewPlugin',
  ).exports
  return ConfigurationSchema(
    'LinearGDCDisplay',
    { renderer: pluginManager.pluggableConfigSchemaType('renderer') },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}
