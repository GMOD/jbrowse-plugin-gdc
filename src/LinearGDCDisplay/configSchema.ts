import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'
import LinearGenomeViewPlugin from '@jbrowse/plugin-linear-genome-view'

export default (pluginManager: PluginManager) => {
  const { baseLinearDisplayConfigSchema } = (
    pluginManager.getPlugin('LinearGenomeViewPlugin') as LinearGenomeViewPlugin
  ).exports
  return ConfigurationSchema(
    'LinearGDCDisplay',
    { renderer: pluginManager.pluggableConfigSchemaType('renderer') },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}
