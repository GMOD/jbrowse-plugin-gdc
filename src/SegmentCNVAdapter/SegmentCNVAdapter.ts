import {
  BaseFeatureDataAdapter,
  BaseOptions,
} from '@jbrowse/core/data_adapters/BaseAdapter'
import { FileLocation, Region } from '@jbrowse/core/util/types'
import { openLocation } from '@jbrowse/core/util/io'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import { SimpleFeature } from '@jbrowse/core/util'
import { Feature } from '@jbrowse/core/util/simpleFeature'
import { readConfObject } from '@jbrowse/core/configuration'
import { AnyConfigurationModel } from '@jbrowse/core/configuration/configurationSchema'
import PluginManager from '@jbrowse/core/PluginManager'
import { getSubAdapterType } from '@jbrowse/core/data_adapters/dataAdapterCache'

export default class SegmentCNVAdapter extends BaseFeatureDataAdapter {
  public static capabilities = ['getFeatures', 'getRefNames']

  public config: any

  private setupP?: Promise<Feature[]>

  public constructor(
    config: AnyConfigurationModel,
    getSubAdapter?: getSubAdapterType,
    pluginManager?: PluginManager,
  ) {
    // @ts-ignore
    super(config, getSubAdapter, pluginManager)
    this.config = config
  }

  private async readSeg() {
    const segLocation = readConfObject(
      this.config,
      'segLocation',
    ) as FileLocation
    const fileContents = (await openLocation(
      segLocation,
      // @ts-ignore
      this.pluginManager,
    ).readFile('utf8')) as string
    const lines = fileContents.split('\n')
    const refNames: string[] = []
    const rows: string[] = []
    let columns: string[] = []
    let refNameColumnIndex = 0
    lines.forEach(line => {
      if (columns.length === 0) {
        columns = line.split('\t')
        const chromosome = (element: any) =>
          element.toLowerCase() === 'chromosome'
        refNameColumnIndex = columns.findIndex(chromosome)
      } else {
        if (line.split('\t')[refNameColumnIndex] !== undefined) {
          rows.push(line)
          refNames.push(line.split('\t')[refNameColumnIndex])
        }
      }
    })

    return {
      lines: rows,
      columns,
      refNames: Array.from(new Set(refNames)),
    }
  }

  private parseLine(line: string, columns: string[]) {
    let segment: any = {}
    line.split('\t').forEach((property: string, i: number) => {
      if (property) {
        if (i === 0) {
          segment.id = property
        } else {
          /* some SEG files have different data, this logic is to ensure that we don't need special
             colouring functions to accomodate for those differences...mean and copy number indicate
             the track colouring */
          if (
            columns[i].toLowerCase() === 'segment_mean' ||
            columns[i].toLowerCase() === 'copy_number'
          ) {
            segment.score = +property
          }
          segment[columns[i].toLowerCase()] = property
        }
      }
    })
    return segment
  }

  private async getLines() {
    const { columns, lines } = await this.readSeg()

    return lines.map((line, index) => {
      const segment = this.parseLine(line, columns)
      return new SimpleFeature({
        uniqueId: segment.id,
        id: segment.id,
        start: +segment.start,
        end: +segment.end,
        refName: segment.chromosome,
        score: +segment.score,
        ...segment,
      })
    })
  }

  private async setup() {
    if (!this.setupP) {
      this.setupP = this.getLines()
    }
    return this.setupP
  }

  public async getRefNames(_: BaseOptions = {}) {
    const { refNames } = await this.readSeg()
    return refNames
  }

  public getFeatures(region: Region, opts: BaseOptions = {}) {
    return ObservableCreate<Feature>(async observer => {
      const feats = await this.setup()
      feats.forEach(f => {
        if (
          f.get('refName') === region.refName &&
          f.get('end') > region.start &&
          f.get('start') < region.end
        ) {
          observer.next(f)
        }
      })
      observer.complete()
    }, opts.signal)
  }

  public freeResources(): void {}
}
