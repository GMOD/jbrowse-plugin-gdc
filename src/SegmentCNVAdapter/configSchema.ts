import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'SegmentCNVAdapter',
  {
    segLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/myfile.seg', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
