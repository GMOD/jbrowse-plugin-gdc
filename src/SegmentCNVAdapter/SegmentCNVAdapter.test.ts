import { toArray } from 'rxjs/operators'
import SegmentCNVAdapter from './SegmentCNVAdapter'
import configSchema from './configSchema'
import fetchMock from 'jest-fetch-mock'

fetchMock.enableMocks()

test('adapter can fetch features from segments.txt', async () => {
  const adapter = new SegmentCNVAdapter(
    configSchema.create({
      segLocation: {
        localPath: require.resolve(
          './test_data/8da2e4c4-e9e2-4af0-9352-3c54e7f4539c.grch38.seg.v2.txt',
        ),
      },
    }),
  )

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: '3',
    start: 0,
    end: 200000,
  })

  const names = await adapter.getRefNames()
  expect(names).toMatchSnapshot()

  const featuresArray = await features.pipe(toArray()).toPromise()
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray).toMatchSnapshot()
})
