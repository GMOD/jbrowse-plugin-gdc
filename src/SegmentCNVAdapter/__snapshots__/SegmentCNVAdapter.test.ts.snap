// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from segments.txt 1`] = `
Array [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "X",
  "Y",
]
`;

exports[`adapter can fetch features from segments.txt 2`] = `
Array [
  Object {
    "end": 137168,
    "id": "0e867d2c-937c-4705-afec-c5aef316c2ba",
    "mean": 0.0633,
    "nprobes": 105,
    "refName": "3",
    "score": 0.0633,
    "start": 20930,
    "uniqueId": "0e867d2c-937c-4705-afec-c5aef316c2ba",
  },
  Object {
    "end": 1247886,
    "id": "0e867d2c-937c-4705-afec-c5aef316c2ba",
    "mean": -0.0747,
    "nprobes": 916,
    "refName": "3",
    "score": -0.0747,
    "start": 137844,
    "uniqueId": "0e867d2c-937c-4705-afec-c5aef316c2ba",
  },
]
`;
